### DISCUSSION TITLE
Forces, Moments, Frames of Reference and Unusual Attitudes

### DISCUSSION THREAD

**Discussion by Murmur79** [JSBSim-Team/jsbsim#512] [Q&A]
Good day everyone,

I have a few questions about the calculation of forces and moments in relation to the frame of reference (FoR), and what happens in some unusual attitudes.

So, I reckon the "wind frame" is the most used FoR in jsbsim flight models (at least, in FlightGear, which is where I come from).

1) First question (see image 1: https://drive.google.com/file/d/1t3DGx4vakmxHcHnXMmWCSCByDo4GLK0C/view?usp=sharing ):

If I use the wind FoR ( `<axis name=”DRAG”>... <axis name=”ROLL”>...`), is it correct to say that the "ROLL", "PITCH", "YAW" calculations in JSBSim (e.g. the definition of aerodynamic coefficients, etc.) refer to the wind FoR (green rotating arrows in image 1) and NOT to the body FoR (red rotating arrows in image 1)?

2) Hence, second question (see image 2: https://drive.google.com/file/d/1sxvOl4T01LR23rkSK8Un6o_by1qCsLDd/view?usp=sharing ):

If alpha = 90 degrees (e.g. during a Cobra maneuver, just imagine the C172 in the image is actually a Su27), is it correct to say that the "ROLL" axis in JSBSim in this case coincides with the YAW body axis, and the "YAW" axis in JSBSim coincides with the ROLL body axis?

In other words, if the <axis name=”ROLL”> definition in JSBSim produces, say a positive rolling moment while alpha = 90 degrees, will it actually cause the aircraft to _YAW_ to the right?

3) I presume that in JSBSim, alpha is defined in the range -180/+180 degrees (or equivalent radians). This should necessarily imply that beta is hence defined in the -90/+90 degrees range, is it correct?

Also, this should imply that IF alpha = +/-180 degrees, YAW and ROLL in BODY AXES are opposite to YAW and ROLL in WIND AXES.

e.g., if alpha = 180 degrees (tailslide), a positive (=nose right in normal attitude) YAW force in JSBSim, would actually induce a "tail right" (and hence a "nose left") rotation of the aircraft. Is this correct?

**Comment by seanmcleod** [comment]
First off JSBSim supports 3 frames of reference for the forces and moments, and the forces and moments don't need to use the same frame of reference.

- Wind 
- Stability
- Body

Before the following commit - https://github.com/JSBSim-Team/jsbsim/commit/eb996220167f38e3a3a46af11a4c31f7bc8360a3 the moments were always required to be in the body frame of reference and the forces could be either wind or body. After this commit all 3 frames of references are available to both forces and moments.

So when you say the wind axes are the most common in terms of FlightGear remember those models are typically using wind axes for the forces but body axes for the moments.

If you look at the code in `FGAerodynamics.cpp` you'll see that if the forces and moments aren't already in the body axes then they're transformed into the body axes before being applied.

https://github.com/JSBSim-Team/jsbsim/blob/ddd025ae2ed76631b4a9b00cc2871a78c251d476/src/models/FGAerodynamics.cpp#L205-L229

In the forces case there is also the `body axial` axes which is the same as the body axes except for a sign difference for the x and z axes.

https://github.com/JSBSim-Team/jsbsim/blob/ddd025ae2ed76631b4a9b00cc2871a78c251d476/src/models/FGAerodynamics.cpp#L261-L273

Then additional moments around the current cg are calculated based on the difference between the MRC (Moment Reference Coordinate) and the current cg and the forces.

https://github.com/JSBSim-Team/jsbsim/blob/ddd025ae2ed76631b4a9b00cc2871a78c251d476/src/models/FGAerodynamics.cpp#L280

And so at this point the sum of the forces are in the body axes and are applied at the cg and the sum of the moments are also in the body axes.

**Comment by Murmur79** [comment]
Thank you for the prompt answer, @seanmcleod 

The situation is still not 100% clear to me. If you don't mind I will need a few other clarifications: :-)

.) I understand the forces and the moments can use two different frames of reference. But I suppose the 3 forces have to use a consistent frame of reference, and the 3 moments a consistent frame of reference (which can be different from the first one), is that correct?

.) What happens if I do not specify a "frame='BODY|STABILITY|WIND'" after the axis definition?
E.g. If I use the LIFT|DRAG|SIDE axis definition, what frame will JSBSim assume?
Similarly, what if I use X|Y|Z or AXIAL|NORMAL|SIDE instead? (without specifying " frame='BODY|STABILITY|WIND' " )

.) Same for the moments: If I use the ROLL|PITCH|YAW without specifying the frame, what frame will JSBSim assume for the moments?

**Comment by seanmcleod** [comment]
> But I suppose the 3 forces have to use a consistent frame of reference, and the 3 moments a consistent frame of reference (which can be different from the first one), is that correct?

Correct, you can't mix and match within the forces or moments axes. You have to choose one for all 3 force axes and one for all 3 moment axes.

> What happens if I do not specify a "frame='BODY|STABILITY|WIND'" after the axis definition?

The `frame` element was a new addition as part of the commit I referenced and is optional. Otherwise every single JSBSim model file would've had to have been updated.

So there is code for handling the case where `frame` isn't specified and figuring out what frame of reference you're specifying based on the name supplied. Which is what JSBSim was doing before, just with a more limited set, i.e. no stability axes available and only body for moments.

**Comment by Murmur79** [comment]
Ok, maybe now it's clearer. Analyzing the code in your link, I understand that:

.) X|Y|Z and ROLL|PITCH|YAW:
-----> default to "body axes";
-----> can be explicitly referenced to either "body", "wind" or "stability axes";

.) LIFT|DRAG|SIDE is (and can only be) referenced to "wind axes";

.) AXIAL|NORMAL|SIDE is (and can only be) referenced to "body axial axes" (which is the "body axes" rotated 180 deg around Y.)

So, if the above is correct:

`<axis name="Z" frame="WIND">` should be equivalent to `<axis name="LIFT">`
`<axis name="X" frame="WIND">` should be equivalent to `<axis name="DRAG">`
`<axis name="Y" frame="WIND">` should be equivalent to `<axis name="SIDE">`

(except possibly for the opposite signs of X and Z axes, if I got the orientation of the body frame correct).

**Comment by seanmcleod** [comment]
Looks right.

One thing I still need to do after that commit added support for all 3 frames of reference for both forces and moments is to update the documentation. You'll notice that the current documentation doesn't mention the new `frame` element option.

https://jsbsim-team.github.io/jsbsim-reference-manual/mypages/user-manual-forces-and-moments/
